### ISSUE TITLE
Using expose_functions() errors when model binary already exists 

### ISSUE THREAD

**Issue by gobbios** [stan-dev/cmdstanr#784]
**Describe the bug**
When using `cmdstan_model("mymod.stan", compile_standalone = TRUE)` and the `mymod` executable already exists (so nothing is (re)compiled) then `$expose_functions()` returns the error:

```
Error in function_env$external && cmdstan_version() < "2.32" :
  invalid 'x' type in 'x && y'
```

If this is not a bug, I'd think that maybe a clearer error message would help.

**To Reproduce**
```
library(cmdstanr)

stancode <- "
  functions {
    real add_up(real a, real b){
    return a + b;
    }
  }
  data {
    int<lower=0> N;
    array[N] int<lower=0,upper=1> y;
  }
  parameters {
    real<lower=0,upper=1> theta;
  }
  model {
    theta ~ beta(1, 1);
    y ~ bernoulli(theta);
  }
"

write_stan_file(stancode, dir = "", basename = "mymod")
```

On first compilation everything runs as expected.

```
mod <- cmdstan_model("mymod.stan", compile_standalone = TRUE)
mod$expose_functions(global = TRUE)
add_up(4, 5)
mod$functions$add_up(4, 5)
```

**Then restart the R session**

```
library(cmdstanr)
mod <- cmdstan_model("mymod.stan", compile_standalone = TRUE)
# Model executable is up to date!

# both lines result in the same error message:
mod$expose_functions(global = TRUE)
mod$expose_functions(global = FALSE)

# Error in function_env$external && cmdstan_version() < "2.32" :
#   invalid 'x' type in 'x && y'

mod$functions$add_up
# NULL

# sampling seems fine
mod$sample(data = list(N = 5, y = c(0, 1, 1, 0, 1)), refresh = 0, chains = 1, seed = 1)

```

**Expected behavior**

I expected that the Stan function would be available (in the global environment, or at least in the model environment) even when recompilation of the model is not necessary.

**Operating system**
macOS Ventura 13.1

**CmdStanR version number**
cmdstanr: 0.5.3 (from `remotes::install_github("stan-dev/cmdstanr")`)
cmdstan_version(): 2.32.2

**Additional context**
none

**Comment by jgabry** [comment]
Thanks for reporting this. I agree there should be a more informative message.  @andrjohns what do you think? Should we add a message saying that recompilation is necessary (if that's the issue here)?

**Comment by jgabry** [comment]
This came up again in https://github.com/stan-dev/cmdstanr/issues/819 from @maxbiostat. @andrjohns What was the intended behavior when the model is already compiled? Should there just be an informative error message saying to recompile? Should it recompile automatically in order to compile the functions? Something else? Thanks!

**Comment by jgabry** [comment]
@andrjohns This error keeps coming up (see e.g. #819 and #830). Do you have time to add an informative error message here? If you don't have time right now no worries, I can try to take care of it, you're just more familiar with the expose_functions code.

**Comment by andrjohns** [comment]
Closed by #831
